### ISSUE TITLE
Decorators should allow a `--` separator

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#9941]
Noticed in #9940:

If you want to decorate something with `command` to only call commands, you can't use a `--` to avoid it being called as e.g. `command -q`:

```fish
command -- less
```

should call `less`, instead it brings up the help for the command *builtin*.

Same goes for `time`, and I have no idea *what* `builtin -- -q` does (it returns 0?).

For variables this already kinda works because of how we parse what - `set -l foo -q grep; command $foo` complains that it can't find the command `-q`.

But regardless, it feels like `--` is the intuitive safeguard and I think it should be added.

**Comment by mqudsi** [comment]
I don't understand why this would be needed; I would say the bug is that `build -- -q` isn't a syntax error or execution error is the real issue and the same with `command -- less` bringing up the help for `builtin`. I would say both of these should just say `-- not found` or whatever in all cases, not just when used with variable substitution and that would resolve your issue with them?

I *guess* an executable *could* be located at `./-a` but that seems *extremely* unlikely and in that case, the `command ./-a` syntax can be used instead of the `command -a` syntax anyway. That option wouldn't work if an executable called `-a` is in `$PATH` but that just seems wrong on so many levels. I guess we totally *can* add `--` support to handle that, but I think the other quirks in how non-existent payloads to `command` are handled should be fixed first in all cases, whether or not `--` separator support is added, and I would hope that `--` never makes an appearance in any of the official documentation or examples and is just there for the weird cases where someone runs into the `$PATH` issue.

More than anything, I just feel like since `command` or `function` is a decorator, it is intended to be tacked on to the name of the head and influences how the parser handles the payload, but it is not conceptually the same as executing `command` and then passing it the `argv` that could be misinterpreted as something else - the handling happens within fish, the parser is aware of it at the core, it's part of the core fish syntax, etc. and it is a decorator in the ast, not something that is executed generically and needs to decide at runtime what behavior should be taken.

**Comment by faho** [comment]
>not just when used with variable substitution and that would resolve your issue with them?

No. The problem is this would be used with `complete`, so the variable substitution would happen *before* the script. Awkward example:

```fish
complete -c foo --condition "command -- $foo" -a "yes no"
```

In this case I want to avoid `$foo` being interpreted as an option.

I feel like `--` is the obvious choice, because it is already used all over the place. I feel that's the consistent solution.

>I guess an executable could be located at ./-a but that seems extremely unlikely and in that case

It could also be at /usr/bin/-a

> I just feel like since command or function is a decorator,

It's a decorator *and a builtin*.

**Comment by lgarron** [comment]
Copying from https://github.com/fish-shell/fish-shell/issues/12998#issuecomment-5649363346 👍 

If you ned a way to way to invoke `command "$BINARY"` so that it works for any valid binary name, use:

```shell
command command -- "$BINARY"
```

This uses the `command` built-in to invoke the `command` binary and relies on its `--` support.

I think it would still be good to support directly in the built-in, as `command command` may work correctly but looks looks like a typo rather than a load-bearing workaround.

**Comment by lgarron** [comment]
@krobelus Thanks for the fix, it works perfectly! 🤩

```shell
# In a `fish-shell` checkout.

set TEMP_DIR (mktemp -d)
printf "#!/usr/bin/env fish\n\necho hi" > $TEMP_DIR/--
chmod +x $TEMP_DIR/--
set PATH $PATH $TEMP_DIR

cargo run -- -c "command -- --"
```

EDIT: also tried some more!

```fish
cd (mktemp -d)
git clone https://github.com/fish-shell/fish-shell
git -C ./fish-shell/ switch --detach 3db2321515cfe4e4c3edba9ce9702507f5db7549
cargo run --manifest-path ./fish-shell/Cargo.toml

# Inside shell
git clone https://codeberg.org/lgarron/edge-case-binaries && cd edge-case-binaries
git switch --detach 3dd8b818cf83424cea51b9367f76359b17596c7c

set PATH $PATH $PWD/bin
for bin_name in (ls ./bin/)
  command -- $bin_name
end
```

```
[--] Binary is running!
[--4+5=9] Binary is running!
[--all] Binary is running!
[--force] Binary is running!
[--help] Binary is running!
[--version] Binary is running!
[-2-3=-5] Binary is running!
[-a] Binary is running!
[-c] Binary is running!
[-e] Binary is running!
[-rf] Binary is running!
[?] Binary is running!
[*] Binary is running!
[\] Binary is running!
[$] Binary is running!
[2+3=5] Binary is running!
```
